### PR TITLE
Vue 58 integrate universal filter ui into filter menu

### PR DESCRIPTION
**Pull request by josafatburmeister**
# Issue: [Vue-58](https://ticketsystem.schul-cloud.org/browse/VUE-58)

## Description

- removed FilterMenu from DataTable and moved it to standalone component FilterMenuStandalone
- started to integrate [vue-filter-ui](https://github.com/adrianjost/vue-filter-ui) into FilterMenuStandalone

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -61,6 +61,7 @@
 		"tiptap": "^1.26.6",
 		"tiptap-extensions": "^1.28.6",
 		"vue": "^2.6.11",
+		"vue-clickout": "^1.0.0",
```
Comment (adrianjost) [inline package.json]: We already have a solution for that with the `v-click-outside` directive.
You can find the implementation [here](https://github.com/schul-cloud/nuxt-client/blob/develop/src/plugins/directives.js#L12).

I think this is doing exactly the same.

File: src/mixins/defaultFilters.unit.js [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+		expect(wrapper.vm.filterTextContains("test", "te")).toBe(true);
+		expect(wrapper.vm.filterTextContains("test", "a")).toBe(false);
+		expect(wrapper.vm.filterTextContains("Test", "te")).toBe(true);
+		expect(wrapper.vm.filterTextContains(11, 1)).toBe(true);
+		expect(wrapper.vm.filterTextContains(11, 2)).toBe(false);
+
+		expect(wrapper.vm.filterTextEquals("test", "test")).toBe(true);
+		expect(wrapper.vm.filterTextEquals("test", "te")).toBe(false);
+		expect(wrapper.vm.filterTextEquals("Test", "test")).toBe(true);
+	});
```
Comment (adrianjost) [inline src/mixins/defaultFilters.unit.js]: Maybe you can use `test.each`/`it.each` to auto generate a lot of tests and group them in a describe called `can filter strings`. Then all the different test cases are independent.

I find this feature of jest really helpfull. You can read more here: https://jestjs.io/docs/en/api#testeachtablename-fn-timeout

File: src/components/organisms/FilterUI/Filter.vue [hunk 3]
```diff
@@ -0,0 +1,379 @@
...
+	props: {
+		labelAdd: { type: String, default: "add filter" },
+		labelApply: { type: String, default: "apply" },
+		labelCancle: { type: String, default: "cancle" },
+		labelRemove: { type: String, default: "remove" },
+		/*
+    "handleUrl": { type: Boolean },
+    "saveState": { type: Boolean },
+		"consistentOrder": {type: Boolean, default: true},
+		*/
```
Comment (adrianjost) [inline src/components/organisms/FilterUI/Filter.vue]:
```suggestion
```

File: src/components/organisms/DataTable/DataTable.vue [hunk 4]
```diff
@@ -230,7 +231,9 @@ export default {
 				// sort strings
 				return a.localeCompare(b);
 			});
-			return sortOrder !== "desc" ? sortedData : sortedData.reverse();
+			return sortOrder !== "desc"
+				? sortedData
+				: sortedData.reverse();
```
Comment (adrianjost) [inline src/components/organisms/DataTable/DataTable.vue]: I think you can safely revert the changes in this file. Especially the `eslint-disable max-lines` does not seem to be necessary.

**Comment by josafatburmeister** [comment]
>     * [ ]  1. Is there a reason why you don't have [npm installed vue-filter-ui](http://docs.vue-filter-ui.surge.sh/1-GettingStarted.html) and instead copied over a bunch of files?
> 
> 
> Where there bugs in the implementation? If so, please let me know and we could fix them directly the original implementation. I can give you write access to the repo as well.
> 
> I propably missed reviewing a lot of code changes but let's start with this and I will continue reviewing.

The main reason for my decision was that the parser most likely needs to be moved from the vue-filter-ui to the FilterMenuStandalone component and therefore the query api of the vue-filter-ui has to be changed. This could imply some bigger changes, but if you agree we can do the changes on the original repo of cause.

File: src/mixins/defaultFilters.unit.js (hunk 2 above)
Comment (josafatburmeister) [reply src/mixins/defaultFilters.unit.js]: Done :+1:

File: package.json (hunk 1 above)
Comment (josafatburmeister) [reply package.json]: If we import vue-filter-ui as npm package we just can remove this again.

Comment (adrianjost) [reply package.json]: Ahh I see. It was just required because you copied over the vue-filter-ui

**Comment by josafatburmeister** [comment]
> > ```
> > * [ ]  1. Is there a reason why you don't have [npm installed vue-filter-ui](http://docs.vue-filter-ui.surge.sh/1-GettingStarted.html) and instead copied over a bunch of files?
> > ```
> > 
> > 
> > Where there bugs in the implementation? If so, please let me know and we could fix them directly the original implementation. I can give you write access to the repo as well.
> > I propably missed reviewing a lot of code changes but let's start with this and I will continue reviewing.
> 
> 
> The main reason for my decision was that the parser most likely needs to be moved from the vue-filter-ui to the FilterMenuStandalone component and therefore the query api of the vue-filter-ui has to be changed. This could imply some bigger changes, but if you agree we can do the changes on the original repo of cause.

The vue-filter-ui package is imported now instead 👍

File: src/components/organisms/FilterUI/Filter.vue (hunk 3 above)
Comment (josafatburmeister) [reply src/components/organisms/FilterUI/Filter.vue]: Not necessary anymore because the due-filter-ui package is imported now.

**Comment by adrianjost** [comment]
@schul-cloud-bot update screenshots

**Comment by adrianjost** [comment]
funktioniert zwar nur so semi gut, aber das fixen wir später in kleineren Tickets :)

**Comment by adrianjost** [comment]
braucht noch paar tests, bevor automerge funktioniert :/ Entweder wir schreiben welche oder mergen manuell
